# This function reads the *.coverage files generated by the samtools depth -b
# command and summarizes coverage per individual and per locus at the sites specified
# in the argument loci. The format for loci is a bed file. bam.dir is the location
# of the *.coverage files and cov.files is vector of complete files names

summarize.depth.at.target.SNP <- function(bam.dir, cov.files, loci){
  cov.list <- lapply(1:length(cov.files), function(i){
    print(i)
    res <- read.table(paste0(bam.dir, "/", cov.files[i]))
    names(res) <- c("locus", "POS", cov.files[i])
    return(res[,c(1,3)])
  })
  
  depth <- select(loci, c(locus, stop))
  for (i in 1:length(cov.list)) {
    depth <- left_join(depth, cov.list[[i]], by = "locus")
  }
  
  # replace NAs with zeroes
  depth[is.na(depth)] <- 0
  depth <- depth[,-2]
  
  # calc mean and median per locus
  depth <- rowwise(depth) %>%
    mutate(mean = mean((c_across(where(is.numeric))))) %>%
    mutate(med = median((c_across(where(is.numeric))))) %>%
    arrange(mean)
  
  # summarize by individual
  depth.per.ind <- data.frame(colMeans(depth[,2:(ncol(depth)-2)])) %>%
    bind_cols(apply(depth[,2:(ncol(depth)-2)], 2, median)) %>% rownames_to_column() 
  depth.per.ind$rowname <- do.call(rbind, lapply(1:nrow(depth.per.ind), function(i){
    strsplit(depth.per.ind$rowname[i], split = ".f")[[1]][1]
  }))
  
  depth.per.ind$gt.10 <- do.call(rbind, lapply(2:(ncol(depth)-2), function(i){
    length(which(depth[,i] > 9))
  }))
  depth.per.ind$gt.20 <- do.call(rbind, lapply(2:(ncol(depth)-2), function(i){
    length(which(depth[,i] > 19))
  }))
  
  names(depth.per.ind) <- c("Sample", "mean", "median","gt.10", "gt.20")
  
  num.inds.genotypable.rd10 <- length(which(depth.per.ind$gt.10 > (length(unique(loci$locus)) * 0.8)))
  num.inds.genotypable.rd20 <- length(which(depth.per.ind$gt.20 > (length(unique(loci$locus)) * 0.8)))
  
  # plot coverage summaries
  line.labels <- data.frame(x = c(20,20), y = c(15,25), label = c("10 reads", "20 reads"))
  
  plots <- list()
  plots$mean.cov <- ggplot(depth) + geom_bar(aes(x = reorder(locus, mean), y = mean), stat = "identity") +
    geom_hline(yintercept = 20) + geom_hline(yintercept = 10) +
    labs(title = project, x = "Ranked loci", y = "Mean Coverage at Target SNP") +
    geom_text(data = line.labels, aes(x = x, y = y, label = label)) +
    theme(axis.text.x=element_blank(), axis.ticks.x=element_blank())
  plots$med.cov <- ggplot(depth) + geom_bar(aes(x = reorder(locus, med), y = med), stat = "identity") +
    geom_hline(yintercept = 20) + geom_hline(yintercept = 10) +
    labs(title = project, x = "Ranked loci", y = "Median Coverage at Target SNP") +
    geom_text(data = line.labels, aes(x = x, y = y, label = label)) +
    theme(axis.text.x=element_blank(), axis.ticks.x=element_blank())
  
  plots$genos.v.mean.depth <-  ggplot(depth.per.ind, aes(x = mean, y = gt.10, col = "blue")) + 
    geom_point() + geom_point(aes(x = mean, y = gt.20, col = "red")) +
    geom_hline(yintercept = 307) + scale_color_manual(labels = c("10", "20"), values = c("blue", "red")) +
    labs(title = paste(project, "\nNum individuals with read depth >= 10 at 80% of loci:", num.inds.genotypable.rd10,
                       "\nNum individuals with read depth >= 20 at 80% of loci:", num.inds.genotypable.rd20),
         x = "Mean Read Depth", y = "Loci Meeting Read Depth Criterion")
  return(list(loc.depth = depth, depth.per.ind = depth.per.ind, plots = plots))
}